### PR TITLE
Dhis2 5971/displaying tabular data and download formatting

### PR DESCRIPTION
**Pull request by HendrikThePendric**
- Adds style to report table
- Fixes filtering, hiding empty rows, and sorting
- Further refactor what once was the TabularReport report and related components.
- Fixes tests for updated files

One point that this PR raises, is the following:
- Currently `parseTableData` is a plain selector
- By adding the `areStatesEqual` option to connect, I am able to reduce the amount of times this selector is being called.
- But it is still called fairly often, i.e. on an selected-org-unit change.
- I think that probably (not sure) for the reporting-rate-summary this is fine.... But if these tables contained a vast volume of rows, I think I'd want to memoize the results.
- I played around with manually caching some results, but decided to revert my changes, because if we would want to memoize selectors we would probably end up using `reselect`....

=> So the question is: do you think we should introduce reselect?

### CODE REVIEW

File: src/components/TabularReport/ReportTable.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+/* React */
+import React, { Fragment } from 'react'
+import PropTypes from 'prop-types'
+
+const Header = (header, index) => <th key={index}>{header}</th>
```
Comment (Mohammer5) [inline src/components/TabularReport/ReportTable.js]: just using a number as a key is dangerous in this case if we have the html report below the form. After requesting a new report and feeding the changes to this component, the keys won't change and it might produce render bugs.

I'm not so sure what `header`, `row` and `cell` are going to be. If it's just a string, you might want to use that as the key, it it's an object with an id, the id would be perfect, like `key={index + cell}` or `key={index + cell.is}`.

File: src/components/TabularReport/styles.css [hunk 2]
```diff
@@ -0,0 +1,33 @@
+.tabular-report table {
+    border-collapse: collapse;
+}
+
+.tabular-report table thead tr {
```
Comment (Mohammer5) [inline src/components/TabularReport/styles.css]: It's not too important in this case, but having multiple parts in a selector slows down rendering performance.
No need to change that here, I guess. But good to keep it in mind that BEM will def. be more performant than this.

File: src/pages/ReportingRateSummary.js [hunk 3]
```diff
@@ -17,22 +17,20 @@ export default class ReportingRateSummary extends React.Component {
...
-                    showBackButton={!!props.reportHtml}
-                    onBackClick={props.unsetReportData}
+                    label={i18n.t('Reporting rate summary')}
                     systemVersion={props.d2.system.version}
                     sectionKey={props.sectionKey}
                 />
                 <Paper style={styles.container}>
                     <Form
-                        showForm={!props.reportHtml}
+                        showForm={true}
```
Comment (Mohammer5) [inline src/pages/ReportingRateSummary.js]: To me this means that `showForm` should not be required and default to true in the Form component.

File: src/redux/actions/reportData.js [hunk 4]
```diff
@@ -45,7 +45,6 @@ export const loadingReportDataError = () => ({
  * @returns {Function}
  */
 export const loadingReportDataStartWithFeedback = () => dispatch => {
```
Comment (Mohammer5) [inline src/redux/actions/reportData.js]: This one should dispatch `loadingReportDataStart` as well. If there's a reason against dispatching the `loadingReportDataStart` action, then this doesn't need to be a thunk, I'd say?

File: src/redux/selectors/blockUi.js [hunk 5]
```diff
@@ -1,5 +1,4 @@
 export const blockUi = state =>
     state.organisationUnits.loading ||
     state.reportPeriod.loading ||
-    state.reportData.loading ||
```
Comment (Mohammer5) [inline src/redux/selectors/blockUi.js]: I think this should def. block the UI state. Users tend to press buttons multiple times which would fire the ajax multiple times (especially if the loading spinner of the report is below the fold)

File: src/redux/selectors/reportingRateSummary/parseTableData.js [hunk 6]
```diff
@@ -1,4 +1,57 @@
...
+const rowIndexesToRead = [
+    1, // Org unit name
+    4, // column values
+    5,
+    6,
+    7,
+    8,
+]
+
 export default function parseTableData(state) {
```
Comment (Mohammer5) [inline src/redux/selectors/reportingRateSummary/parseTableData.js]: Selectors should always start with `get` (also part of the [result of the discussion](https://github.com/dhis2/notes/pull/19) we had last week friday and yesterday). Also `parse` has a different meaning than `build` or sth like that, so it might be even more confusing to use `parse` here (as well as for the other selectors).

File: src/components/TabularReport/ReportTable.js (hunk 1 above)
Comment (HendrikThePendric) [reply src/components/TabularReport/ReportTable.js]: I'll change this to be more compliant, but I don't it matters in this case. The table follows this render cycle:
1. Render nothing
2. Render loader
3. Render table
4. Render loader
5. Render table

At no point in time are we replacing rows. Keys are important when children are dynamic. Our children are not.

https://stackoverflow.com/questions/42801343/what-is-the-significance-of-keys-in-reactjs/42801409
> TL;DR Use unique and constant keys when rendering dynamic children, or expect strange things to happen.

File: src/pages/ReportingRateSummary.js (hunk 3 above)
Comment (HendrikThePendric) [reply src/pages/ReportingRateSummary.js]: Aha yes, I actually did that for another component but forgot it here.

File: src/redux/actions/reportData.js (hunk 4 above)
Comment (HendrikThePendric) [reply src/redux/actions/reportData.js]: Good one. I removed this to stop showing the spinner overlay, because the table has its own loader. But I shouldn't have changed this thunk. Instead I should have updated the `loadReportData` thunk in `./redux/actions/reportingRateSummary.js`. Updated this now.

File: src/redux/selectors/blockUi.js (hunk 5 above)
Comment (HendrikThePendric) [reply src/redux/selectors/blockUi.js]: I disagree, and will keep things this way. In most DHIS2 apps we show an inline loader for things like lists and tables. I would like to keep it that way.

Comment (Mohammer5) [reply src/redux/selectors/blockUi.js]: This would make the report being loaded many times if the report is loaded below the screen and the user clicks the button. Then the submit btn needs to be disabled during loading.

Comment (HendrikThePendric) [reply src/redux/selectors/blockUi.js]: Done
